src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -0,0 +1,238 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import static java.nio.file.LinkOption.NOFOLLOW_LINKS;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+
+import build.buildfarm.instance.Instance;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Queues;
+import com.google.devtools.remoteexecution.v1test.Digest;
+import io.grpc.StatusRuntimeException;
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.PosixFileAttributes;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+class CASFileCache {
+  private final Instance instance;
+  private final Path root;
+  private final long maxSizeInBytes;
+  private final Map<Path, FileEntry> paths;
+  private final List<Path> accessScanPaths;
+
+  private long sizeInBytes;
+  private Queue<FileEntry> expirationQueue;
+
+  private static final class FileEntry implements Comparable<FileEntry> {
+    private final Path path;
+    private final long sizeInBytes;
+    private final long createdTime;
+    private long accessTime;
+
+    public FileEntry(
+        Path path, long sizeInBytes, long createdTime, long accessTime) {
+      this.path = path;
+      this.sizeInBytes = sizeInBytes;
+      this.createdTime = createdTime;
+      this.accessTime = accessTime;
+    }
+
+    public Path getPath() {
+      return path;
+    }
+
+    public long getSizeInBytes() {
+      return sizeInBytes;
+    }
+
+    public long getCreatedTime() {
+      return createdTime;
+    }
+
+    public long getAccessTime() {
+      return accessTime;
+    }
+
+    public void setAccessTime( long accessTime ) {
+      this.accessTime = accessTime;
+    }
+
+    @Override
+    public int compareTo(FileEntry entry) {
+      // the most recently accessed file is greater
+      if (accessTime != entry.accessTime) {
+        return entry.accessTime > accessTime ? 1 : -1;
+      }
+      // the smaller file is greater
+      if (sizeInBytes != entry.sizeInBytes) {
```

Comment:
Is this and the stuff below necessary? If the access time is sufficiently fine-grained, then this is never executed, i.e., dead code.

Reply:
Dropped the access time stuff entirely